src/python/pants/backend/python/macros/poetry_requirements.py

```diff
@@ -0,0 +1,287 @@
+            (
+                "No requirements defined in poetry.tools.dependencies and"
+                f" poetry.tools.dev-dependencies in {file_path}, which is loaded by Pants"
+                " from a poetry_requirements macro. Did you mean to populate these"
+                " with requirements?"
+            )
+        )
+
+    return set(
+        itertools.chain.from_iterable(
+            parse_single_dependency(proj, attr)
+            for proj, attr in {**dependencies, **dev_dependencies}.items()
+        )
+    )
+
+
+class PoetryRequirements:
+    """Translates dependencies specified in a  pyproject.toml Poetry file to a set of
+    "python_requirements_library" targets.
+
+    For example, if pyproject.toml contains the following entries under
+    poetry.tool.dependencies: `foo = ">1"` and `bar = ">2.4"`,
+
+    python_requirement_library(
+        name="foo",
+        requirements=["foo>1"],
+      )
+
+      python_requirement_library(
+        name="bar",
+        requirements=["bar>2.4"],
+      )
+
+    See Poetry documentation for correct specification of pyproject.toml:
+    https://python-poetry.org/docs/pyproject/
+
+    You may also use the parameter `module_mapping` to teach Pants what modules each of your
+    requirements provide. For any requirement unspecified, Pants will default to the name of the
+    requirement. This setting is important for Pants to know how to convert your import
+    statements back into your dependencies. For example:
+
+        python_requirements(
+          module_mapping={
+            "ansicolors": ["colors"],
+            "setuptools": ["pkg_resources"],
+          }
+        )
+    """
+
+    def __init__(self, parse_context):
+        self._parse_context = parse_context
+
+    def __call__(
+        self,
+        pyproject_toml_relpath: str = "pyproject.toml",
+        *,
+        module_mapping: Optional[Mapping[str, Iterable[str]]] = None,
+    ) -> None:
+        """
+        :param pyproject_toml_relpath: The relpath from this BUILD file to the requirements file.
+            Defaults to a `requirements.txt` file sibling to the BUILD file.
+        :param module_mapping: a mapping of requirement names to a list of the modules they provide.
+            For example, `{"ansicolors": ["colors"]}`. Any unspecified requirements will use the
+            requirement name as the default module, e.g. "Django" will default to
+            `modules=["django"]`.
+        """
+        req_file_tgt = self._parse_context.create_object(
+            "_python_requirements_file",
+            name=pyproject_toml_relpath.replace(os.path.sep, "_"),
+            sources=[pyproject_toml_relpath],
+        )
+        requirements_dep = f":{req_file_tgt.name}"
+
+        req_file = Path(get_buildroot(), self._parse_context.rel_path, pyproject_toml_relpath)
+        requirements = parse_pyproject_toml(
+            req_file.read_text(), str(req_file.relative_to(get_buildroot()))
+        )
```

Comment:
As with `python_requirements`, this macro will suffer from #7022. We have a hacky workaround in place for that issue via #9946: we'll need to do the same for `pyproject.toml`, in this PR.

Reply:
Spent a bit examining updated code; couldn't find code that seems to hack a fix in a similar way. Was this fix moved to Rust code? If not, would you mind linking the file that would be updated to resolve the error?

Reply:
It now lives here: https://github.com/pantsbuild/pants/blob/8966adfd67a43e583254dae0f405cb2d466f5098/src/python/pants/option/global_options.py#L1511-L1521

Reply:
Awesome -- thanks. Fixed in latest commit!